numpy/_core/src/multiarray/unique.cpp

```diff
@@ -0,0 +1,189 @@
+    Py_INCREF(descr);
+    return PyArray_NewFromDescr(
+        &PyArray_Type,
+        descr,
+        1, // ndim
+        dims, // shape
+        NULL, // strides
+        res, // data
+        NPY_ARRAY_OUT_ARRAY, // flags
+        NULL // obj
+    );
+}
+
+
+// this map contains the functions used for each item size.
+typedef std::function<PyObject *(PyArrayObject *)> function_type;
+std::unordered_map<int, function_type> unique_funcs = {
+    {NPY_BYTE, unique<npy_byte>},
+    {NPY_UBYTE, unique<npy_ubyte>},
+    {NPY_SHORT, unique<npy_short>},
+    {NPY_USHORT, unique<npy_ushort>},
+    {NPY_INT, unique<npy_int>},
+    {NPY_UINT, unique<npy_uint>},
+    {NPY_LONG, unique<npy_long>},
+    {NPY_ULONG, unique<npy_ulong>},
+    {NPY_LONGLONG, unique<npy_longlong>},
+    {NPY_ULONGLONG, unique<npy_ulonglong>},
+    {NPY_INT8, unique<npy_int8>},
+    {NPY_INT16, unique<npy_int16>},
+    {NPY_INT32, unique<npy_int32>},
+    {NPY_INT64, unique<npy_int64>},
+    {NPY_UINT8, unique<npy_uint8>},
+    {NPY_UINT16, unique<npy_uint16>},
+    {NPY_UINT32, unique<npy_uint32>},
+    {NPY_UINT64, unique<npy_uint64>},
+};
+
+
+extern "C" NPY_NO_EXPORT PyObject *
+PyArray_Unique(PyObject *NPY_UNUSED(dummy), PyObject *args)
+{
+    /* This is called from Python space, and expects a single numpy array as input.
+
+    It then returns a numpy array containing the unique values of the input array.
+
+    If the input array is not supported, it returns None.
+    */
+    PyArrayObject *self = NULL;
+    PyObject *res = NULL;
+    if (!PyArg_ParseTuple(args, "O&", PyArray_Converter, &self))
+        return NULL;
+    
+    // Making sure the DECREF is called when the function returns, with
+    // or w/o an exception
+    auto self_decref = finally([&]() { Py_XDECREF(self); });
+
+
+    /* Handle zero-sized arrays specially */
+    if (PyArray_SIZE(self) == 0) {
+        return PyArray_NewLikeArray(
+            self,
+            NPY_ANYORDER,
+            NULL, // descr (use prototype's descr)
+            1 // subok
+        );
+    }
+
+    auto type = PyArray_TYPE(self);
+    // we only support data types present in our unique_funcs map
+    if (unique_funcs.find(type) == unique_funcs.end()) {
+        return Py_None;
+    }
+
+    res = unique_funcs[type](self);
+    return res;
+}
+
+
+
+// The following is to expose the unique function to Python
+static PyMethodDef UniqueMethods[] = {
+    {"unique_hash",  PyArray_Unique, METH_VARARGS,
+     "Collect unique values via a hash map."},
+    {NULL, NULL, 0, NULL}
+};
+
+static struct PyModuleDef uniquemodule = {
+    PyModuleDef_HEAD_INIT,
+    "_unique", /* name of module */
+    NULL, /* module docs */
+    -1,  /* size of per-interpreter state of the module,
+            or -1 if the module keeps state in global variables. */
+    UniqueMethods
+};
+
+PyMODINIT_FUNC
+PyInit__unique(void)
+{
+    PyArray_ImportNumPyAPI();
+    return PyModule_Create(&uniquemodule);
+}
```

Comment:
I feel like we should be able to just include this into `multiarray.c` and add the private helper function there (avoiding the additional module)?